### PR TITLE
wal-verify update and refactor

### PR DESCRIPTION
**Pull request by usernamedt**
In this PR I propose to:

- refactor the `wal-verify` command to support running multiple checks against the WAL storage folder
- move the existing command behavior to `integrity` check
- introduce a new `timeline` check. 

`timeline` check verifies if the current cluster timeline is greater than or equal to any of the storage WAL segments timelines. This check is useful to detect split-brain conflicts. 

This check will correctly work only if new storage created, or the existing one cleaned when restoring from the backup or performing `pg_upgrade`.

The output consists of:
1. Status of `timeline` check:
    * `OK` if current timeline id matches the highest timeline id found in storage
    * `WARNING` if could not determine if current timeline matches the highest in storage
    * `FAILURE` if current timeline id is not equal to the highest timeline id found in storage
2. Current timeline id.
3. The highest timeline id found in WAL storage folder.

`wal-verify` usage examples:
``` bash
wal-g wal-verify [space separated list of checks]
# For example:
wal-g wal-verify integrity timeline # perform integrity and timeline checks
wal-g wal-verify integrity # perform only integrity check
```
Command output has been also redesigned a little bit to support future extensions.

Example of the new plain text output:
```
xxx /home/usernamedt $ ./wal-g wal-verify integrity timeline  --config /etc/wal-g/wal-g.yaml 
INFO: 2020/12/07 19:40:13.414012 Building check runner: integrity
INFO: 2020/12/07 19:40:14.181305 Detected earliest available backup: base_00000003000000030000004D
INFO: 2020/12/07 19:40:14.425737 Running the check: integrity
INFO: 2020/12/07 19:40:14.426958 Building check runner: timeline
INFO: 2020/12/07 19:40:14.604540 Running the check: timeline
[wal-verify] integrity check status: OK
[wal-verify] integrity check details:
+-----+--------------------------+--------------------------+----------------+--------+
| TLI | START                    | END                      | SEGMENTS COUNT | STATUS |
+-----+--------------------------+--------------------------+----------------+--------+
|   3 | 00000003000000030000004D | 0000000300000004000000F0 |            420 |  FOUND |
|   4 | 0000000400000004000000F1 | 000000040000000800000034 |            836 |  FOUND |
+-----+--------------------------+--------------------------+----------------+--------+
[wal-verify] timeline check status: OK
[wal-verify] timeline check details:
Highest timeline found in storage: 4
Current cluster timeline: 4
```

Example of the new JSON output:
```
{
   "integrity":{
      "status":"OK",
      "details":[
         {
            "timeline_id":3,
            "start_segment":"00000003000000030000004D",
            "end_segment":"0000000300000004000000F0",
            "segments_count":420,
            "status":"FOUND"
         },
         {
            "timeline_id":4,
            "start_segment":"0000000400000004000000F1",
            "end_segment":"000000040000000800000034",
            "segments_count":836,
            "status":"FOUND"
         }
      ]
   },
   "timeline":{
      "status":"OK",
      "details":{
         "current_timeline_id":4,
         "highest_storage_timeline_id":4
      }
   }
}
```

### CODE REVIEW

File: internal/integrity_check_runner.go [hunk 1]
```diff
@@ -0,0 +1,298 @@
...
+
+func NewIntegrityCheckRunner(rootFolder storage.Folder, currentWalSegment WalSegmentDescription) (IntegrityCheckRunner, error) {
+	walFolder := rootFolder.GetSubFolder(utility.WalPath)
+
+	timelineSwitchMap, err := createTimelineSwitchMap(currentWalSegment.Timeline, walFolder)
+	if err != nil {
+		return IntegrityCheckRunner{}, errors.Wrap(err, "Failed to initialize timeline history map")
+	}
+
+	stopWalSegmentNo, err := getEarliestBackupStartSegmentNo(timelineSwitchMap, currentWalSegment.Timeline, rootFolder)
```
Comment (g0djan) [inline internal/integrity_check_runner.go:30]: I guess the logic is the same as was in wal_handler.go? And now you check the earliest backup also.

File: internal/timeline_check_runner.go [hunk 2]
```diff
@@ -0,0 +1,98 @@
...
+// StatusFailure if current timeline is not equal to the highest timeline id found in storage
+func newTimelineCheckResult(currentTimeline, highestTimeline uint32) WalVerifyCheckResult {
+	result := WalVerifyCheckResult{
+		Status: StatusWarning,
+		Details: TimelineCheckResult{
+			CurrentTimelineId:        currentTimeline,
+			HighestStorageTimelineId: highestTimeline,
+		},
+	}
+	if highestTimeline > 0 {
```
Comment (g0djan) [inline internal/timeline_check_runner.go:55]: We get a warning when highestTimeline == 0. Is it supposed?

File: internal/timeline_check_runner.go [hunk 3]
```diff
@@ -0,0 +1,98 @@
...
+		if currentTimeline == highestTimeline {
+			result.Status = StatusOk
+		} else {
+			result.Status = StatusFailure
+		}
+	}
+	return result
+}
+
+// TODO: Unit tests
```
Comment (g0djan) [inline internal/timeline_check_runner.go:65]: Do it! :)

**Review by g0djan — CHANGES_REQUESTED** [review]

File: internal/integrity_check_runner.go (hunk 1 above)
Comment (usernamedt) [reply internal/integrity_check_runner.go:30]: yes

File: internal/timeline_check_runner.go (hunk 2 above)
Comment (usernamedt) [reply internal/timeline_check_runner.go:55]: yes, highest timeline == 0 means that something is not right. For example, there are no WAL segments in storage

File: PostgreSQL.md [hunk 4]
```diff
@@ -233,27 +233,42 @@ By default, `wal-show` output is plaintext table. For detailed JSON output, add
...
+3. The highest timeline id found in WAL storage folder.
+
+Usage:
+``` bash
+wal-g wal-verify [space separated list of checks]
+# For example:
+wal-g wal-verify integrity timeline # perform integrity and timeline checks
+wal-g wal-verify integrity # perform only integrity check
+```
+
```
[the remark this answers is not in the document]
Comment (usernamedt) [reply PostgreSQL.md:271]: Fixed in 7d1e954

File: cmd/pg/wal_verify.go [hunk 5]
```diff
@@ -10,35 +13,71 @@ import (
...
+		checkType, ok := availableChecks[check]
+		if !ok {
+			tracelog.ErrorLogger.Fatalf("Check %s is not available.", check)
+		}
+		checkTypes = append(checkTypes, checkType)
+	}
+	return checkTypes
+}
+
+func checkArgs(cmd *cobra.Command, args []string) error {
```
[the remark this answers is not in the document]
Comment (usernamedt) [reply cmd/pg/wal_verify.go:70]: Fixed in 6f59e57

File: internal/wal_verify_handler.go [hunk 6]
```diff
@@ -1,76 +1,68 @@
...
-	// Storage contains lost missing segments
-	Failure
+	WalVerifyIntegrityCheck = iota + 1
+	WalVerifyTimelineCheck
 )
 
-func (storageStatus WalStorageStatus) String() string {
-	return [...]string{"", "OK", "WARNING", "FAILURE"}[storageStatus]
+func (status WalVerifyCheckType) String() string {
+	return [...]string{"", "integrity", "timeline"}[status]
```
[the remark this answers is not in the document]
Comment (usernamedt) [reply internal/wal_verify_handler.go]: Fixed in 9e13781

File: internal/timeline.go [hunk 7]
```diff
@@ -73,8 +73,13 @@ var (
 	// WalSegmentSize is the size of one WAL file
 	WalSegmentSize        = uint64(16 * 1024 * 1024)
 	xLogSegmentsPerXLogId = 0x100000000 / WalSegmentSize // xlog_internal.h line 101
+	walHistoryFileRegexp  *regexp.Regexp
 )
 
+func init() {
+	walHistoryFileRegexp = regexp.MustCompile("^([0-9a-fA-F]+)\\.history(\\.\\w+)?$")
```
[the remark this answers is not in the document]
Comment (usernamedt) [reply internal/timeline.go]: Fixed in 360c202

File: internal/timeline_check_runner.go (hunk 3 above)
Comment (usernamedt) [reply internal/timeline_check_runner.go:73]: I've added wal-verify tests in 861313e

File: internal/wal_verify_handler_test.go [hunk 8]
```diff
@@ -0,0 +1,540 @@
...
+}
+
+// check that wal-verify correctly follows timeline switches
+func TestWalVerify_TwoTimelines_Ok(t *testing.T) {
+	storageSegments := []string{
+		"000000050000000000000001",
+		"000000050000000000000002",
+		"000000050000000000000003",
+		"000000050000000000000004",
+		"000000050000000000000005", // should not get into output
```
Comment (g0djan) [inline internal/wal_verify_handler_test.go:227]: @x4m take a look. Is it okay if we have some more WAL's on the previous timeline in the storage?

File: internal/wal_verify_handler_test.go [hunk 9]
```diff
@@ -0,0 +1,540 @@
...
+				EndSegment:    "000000050000000000000002",
+				SegmentsCount: 2,
+				Status:        internal.Found,
+			},
+			{
+				TimelineId:    5,
+				StartSegment:  "000000050000000000000003",
+				EndSegment:    "000000050000000000000003",
+				SegmentsCount: 1,
+				Status:        internal.Lost,
```
Comment (g0djan) [inline internal/wal_verify_handler_test.go:324]: Missing WAL's on an old timeline is "Lost".

File: internal/wal_verify_handler_test.go [hunk 10]
```diff
@@ -0,0 +1,540 @@
...
+				EndSegment:    "000000050000000000000004",
+				SegmentsCount: 1,
+				Status:        internal.Found,
+			},
+			{
+				TimelineId:    6,
+				StartSegment:  "000000060000000000000005",
+				EndSegment:    "000000060000000000000006",
+				SegmentsCount: 2,
+				Status:        internal.ProbablyUploading,
```
Comment (g0djan) [inline internal/wal_verify_handler_test.go:338]: And missing WAL's on the current timeline is always considered as "ProbablyUploading"?

File: internal/wal_verify_handler_test.go [hunk 11]
```diff
@@ -0,0 +1,540 @@
...
+		"000000060000000000000007",
+		"000000060000000000000008",
+	}
+
+	storageFiles := make(map[string]*bytes.Buffer, 4)
+
+	backupSentinelNames := []string{
+		// this backup should not be selected as the earliest,
+		// because it does not belong to the current timeline history
+		utility.BackupNamePrefix + "000000050000000000000005" + utility.SentinelSuffix,
```
Comment (g0djan) [inline internal/wal_verify_handler_test.go:433]: But it's the earliest backup

**Review by g0djan — CHANGES_REQUESTED** [review]
Good job 👍.

File: internal/wal_show_handler_test.go [hunk 12]
```diff
@@ -47,12 +47,12 @@ func (timelineSetup *TestTimelineSetup) GetWalFilenames() []string {
 	return filenamesWithExtension
 }
 
-// GetHistory returns .history file name and compressed contents
-func (timelineSetup *TestTimelineSetup) GetHistory() (string, *bytes.Buffer, error) {
+// newTimelineHistoryFile returns .history file name and compressed contents
+func newTimelineHistoryFile(contents string, timelineId uint32) (string, *bytes.Buffer, error) {
```
[the remark this answers is not in the document]
Comment (usernamedt) [reply internal/wal_show_handler_test.go:51]: https://jakubmarian.com/content-vs-contents-in-english/

In the plural, “contents” refer to the things contained in something:

In a concrete sense. We can speak of the “contents of a bag”, “contents of a bottle”, “contents of the stomach”, etc., for instance,
He dropped a bottle and spilled its contents all over the floor.
In an abstract sense. Traditionally, pieces of information contained in a written document are called its “contents”. You may say,
She has lost the letter and remained unaware of its contents.

So I think it is legit

[the remark this answers is not in the document]
Comment (g0djan) [reply internal/wal_show_handler_test.go:51]: Fine)

File: internal/wal_verify_handler_test.go [hunk 13]
```diff
@@ -0,0 +1,540 @@
...
+
+func init() {
+	// Set upload disk concurrency to non-zero value
+	// Please note: this setting affects wal-verify behavior
+	viper.Set(internal.UploadConcurrencySetting, "4")
+}
+
+type WalVerifyTestSetup struct {
+	expectedIntegrityCheck internal.WalVerifyCheckResult
+	expectedTimelineCheck internal.WalVerifyCheckResult
```
[the remark this answers is not in the document]
Comment (usernamedt) [reply internal/wal_verify_handler_test.go]: Fixed in 5458da5

File: internal/wal_verify_handler_test.go (hunk 8 above)
Comment (usernamedt) [reply internal/wal_verify_handler_test.go]: I've added some comments to clarify this in 27a4884

File: internal/wal_verify_handler_test.go (hunk 10 above)
Comment (usernamedt) [reply internal/wal_verify_handler_test.go:442]: I've added illustration to docs in a1fd3b3

File: internal/wal_verify_handler_test.go (hunk 11 above)
Comment (usernamedt) [reply internal/wal_verify_handler_test.go:546]: Clarified this in 27a4884

File: internal/wal_verify_handler_test.go (hunk 9 above)
Comment (usernamedt) [reply internal/wal_verify_handler_test.go:428]: Lost segments may be on any timeline

File: internal/wal_verify_handler_test.go [hunk 14]
```diff
@@ -3,30 +3,28 @@ package internal_test
...
 	"github.com/wal-g/wal-g/internal"
 	"github.com/wal-g/wal-g/utility"
 	"reflect"
 	"testing"
 )
 
-func init() {
-	// Set upload disk concurrency to non-zero value
-	// Please note: this setting affects wal-verify behavior
-	viper.Set(internal.UploadConcurrencySetting, "4")
```
Comment (g0djan) [inline internal/wal_verify_handler_test.go]: why did you remove this line?

**Review by g0djan — APPROVED** [review]

File: internal/wal_verify_handler_test.go (hunk 14 above)
Comment (usernamedt) [reply internal/wal_verify_handler_test.go]: Because I've added this below:
```
// Please note: this settings affect wal-verify behavior
 // Read more in PostgreSQL.md
 const uploadingSegmentRangeSize = 4
 const delayedSegmentRangeSize = 3
```

Comment (usernamedt) [reply internal/wal_verify_handler_test.go]: This was designed to make the logic more straightforward, but it looks like it did the opposite.

Reverted in 6827203
